Logo analysis

MCharming98 opened the pull request:
Implementation of server-side image analysis part.

maksymko commented on src/main/java/com/google/codeu/servlets/LogoDetectServlet.java:
```diff
@@ -0,0 +1,109 @@
+      byte[] blobBytes = getBlobBytes(blobstoreService, blobKey);
+      String logos = getLogoInformation(blobBytes);
+      String[] logosArray = logos.split(",");
+      this.logoJsonArray.add(logosArray[0]); //Chose the first logo for now
```
I think I see now where the main confusion is. Your logoJsonArray is a global object that exists on a servlet level. The servlet can potentially serve many requests at the same time, so there is no way for you to guarantee that the user calling doGet (indirectly) of this servlet is the same user who earlier called "doPost". You need to find a way to pass this information to `/info-present.html`, which does not involve using the servlet object data.

maksymko commented on src/main/webapp/js/wikiSearch.js:
```diff
@@ -0,0 +1,38 @@
+function searchOnWiki(){
+  fetch('/logo-detect').then(function(response) {
+    return response.json();
```
Over here you assume that the JSON in the response will correspond to the request that you've made before. There is no way to guarantee that. If you test it on your local machine, where you have only one serving machine and only one user (you) it *will* work. However, if you run it on AppEngine, where the servlet will run on multiple machines the `logoJsonArray` will be private data, different on each serving machine.

It's quite possible that your request to upload the image and the response to /logo-detect will be served by different machines (or processes) and each of those will have different value for that parameter.

maksymko commented on src/main/java/com/google/codeu/servlets/LogoDetectServlet.java:
```diff
@@ -1,32 +1,119 @@
 public class LogoDetectServlet extends HttpServlet {
 
-  JsonArray logoJsonArray;
+  private String keyword;
```
Why do you still need this?

maksymko commented on src/main/java/com/google/codeu/servlets/LogoDetectServlet.java:
```diff
@@ -1,32 +1,119 @@
+      String[] logosArray = logos.split(",");
+      this.keyword = logosArray[0]; //Chose the first logo for now
+    }
+    response.sendRedirect("/info-present.html?kw=" + keyword);
```
Looks like the local variable will do just fine.

MCharming98 committed:
Add submit button to connect to servlet

MCharming98 committed:
Add error handling in wikiSearch.js

MCharming98 committed:
Finalize connecting logo recognition with image uploading

MCharming98 committed:
Optimized the result from vision API to avoid querying wikipedia a vague term like "mini starbucks".
Added options list when the api returns a vague result.
Improved presentation of image and result.

maksymko commented on src/main/java/com/google/codeu/servlets/LogoDetectServlet.java:
```diff
@@ -25,6 +34,83 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
+        List<String> logos = detectLogos(blobBytes);
+        String keyWord = logos.get(0); //Chose the first logo for now
+        response.sendRedirect("/info-present.html?kw=" + keyWord);
+      } catch (Exception e) {
```
This is pretty broad and at least requires some explanation (i.e. comment). If there is a specific error `detectLogos` throws that you want to catch -- then catch that specific error. In any case you need to log the error somewhere, so that you can look at it.

maksymko replied:
Actually, I would suggest removing the catch and let the error bubble up. If it becomes frequent and annoying, then it needs to be addressed, but at that time you would at least know what the problem is. If the exception just silently ignored, it will just appear unstable for no particular reason.

maksymko commented on src/main/webapp/js/home.js:
```diff
@@ -19,7 +19,7 @@ function displayImage() {
       var img = document.querySelector('img');
       img.src = URL.createObjectURL(this.files[0]);
-      img.onload = imageIsLoaded;
+      //img.onload = imageIsLoaded;
```
Just delete the line if you no longer need it.

MCharming98 replied:
@alyssa-lau This line causes an undefined variable error, and removing it seems not affect the image upload process, is it good to delete this line?

alyssa-lau replied:
Yes, we can delete this line.
